### ISSUE TITLE
NVDA incorrectly identifies word boundaries in Java applications

### ISSUE THREAD

**Issue by kranid** [nvaccess/nvda#16374]
### Steps to reproduce:
1. Install IntelliJ IDEA 2024.1 or higher https://www.jetbrains.com/idea/download/
2. Run the Jaccess inspector and select "Track caret property events" checkbox  in "Accessibility events" menu
3. Open any symple java project for example: https://github.com/kranid/test
4. Open Main.java in the code editor
5. Ensure that the "When moving by words" option in settings/editor/general is set to the default value, which is "Jump to the current word boundaries."
6. Press ctrl+g, type "4,13" and press enter to  Set the keyboard focus to the first letter of the line: "System.out.print(i);"
7. Use Ctrl+Right Arrow to move by words
8. ensure that jaccess inspector correctly identifies  the words but NVDA does not.
### Actual behavior:

NVDA attempts to identify word boundaries using its own algorithms instead of utilizing the Java Accessibility API to retrieve words provided by the Java application so NVDA and IDEA consider as separate words is various text segments .

|caret index| Character|Jaccess inspector|NVDA|
|---|----|-----|--------|
|120|.|System|System.out.print|
|121|o|.|System.out.print|
|124|.|out|System.out.print|
|125|p|.|System.out.print|
|130|(|print|(i);|

### Expected behavior:

NVDA correctly identifies word boundaries in Java applications
### NVDA logs, crash dumps and other attachments:
[jaccessinspector.log](https://github.com/nvaccess/nvda/files/14911710/jaccessinspector.log)

#### NVDA installed/portable/running from source:
installed
#### NVDA version:
2024.1
#### Windows version:
Edition	Windows 11 Pro
Version	23H2
OS build	22631.3374

#### Name and version of other software in use when reproducing the issue:
IntelliJ IDEA 2024.1 (Community Edition)
Build #IC-241.14494.240, built on March 28, 2024
Runtime version: 17.0.10+8-b1207.12 amd64
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
The behavior does not depend on NVDA version.
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by seanbudd** [comment]
Potentially related #16237

**Comment by seanbudd** [comment]
cc @michaelDCurran @mwhapples

**Comment by mwhapples** [comment]
Performing testing with a Braille display I make the following observations. I get similar speech output to that described in this issue. On the Braille display the cursor is updated correctly. So it seems like NVDA is getting the correct caret movement events. My thought though is what precisely should the output be? Looking at the information given in the actual section and how it compares with jaccess inspector and then comparing with similar actions in notepad with NVDA, I am left with multiple options and no definite feeling of which is the "correct" behaviour. It appears the output for jaccess in this issue relates to the text between the previous and current caret position (IE. the text moved over). In notepad NVDA speaks the word the cursor lands on (eg. System . out . println). So why in IntelliJ it speaks more than just the word (IE. System.out.println) I don't know why there is this difference.

**Comment by mwhapples** [comment]
As my previous comment contained a lot of information, for clarity here is my question: Should NVDA speak the word it lands on or the text moved over by the cursor movement event? The former matches NVDA in notepad, the latter matches what has been reported for jaccess inspector.

**Comment by kranid** [comment]
@mwhapples, I believe IntelliJ IDEA should define what constitutes a word, at least until NVDA does not manage the cursor and delegates this function to the current application. Users can adjust cursor behavior using IntelliJ IDEA settings to modify word definition. Therefore, I assume NVDA simply needs to announce the words provided by IntelliJ IDEA through the Java Accessibility API; so i sure the JAccess Inspector shows the truly correct words.

**Comment by mwhapples** [comment]
The thing is that what I observe in jaccess inspector does not seem to report the word I would expect. The problem I observe with jaccess inspector is that if the cursor is placed at the start of a word or may be more precisely put between words, then whilst the current character is reported as the character after the cursor (ie. the one which would be deleted if the user were to press the delete key), the current word is being reported as the word before the cursor (IE. what would be removed when using the backspace key). Thus the reported current character is not part of the reported current word and so it does not seem right to rely upon this part of the API in this case. I need to test whether this is a bug in IntelliJ and what it passes through java access bridge or whether its a deeper java issue (may be access bridge or may be a Java UI toolkit thing).
This bug in what is reported by access bridge does not show itself when the cursor is mid word. However the bug is significant as when navigating by word the cursor is placed at the start of a word and so in a position where the bug is present.

**Comment by mwhapples** [comment]
So for this case there does seem to be a bug in IntelliJ not reporting the correct current word when the cursor is at the start of the word. The Java notepad demmo app in the JDK reports what I would have expected, so it does seem to be an IntelliJ specific issue and I will report a bug there. Even when/if that is fixed then NVDA would need to be modified to use this information from the (java Access Bridge.

**Comment by kranid** [comment]
@mwhapples, I'd like to draw your attention to the fact that by default, the cursor is positioned at the end rather than the start of a word. It is placed between words or at the start of the next word, but IDEA considers this position as the end of the previous word. For instance, if we take the line "    void main()", and the cursor is initially set to the first character of the line, which is a space, pressing Ctrl+Right will move the cursor between "void" and "main". Subsequent Ctrl+Right presses will position the cursor at the left parenthesis. It's worth noting that the left parenthesis is considered a word, meaning the cursor is set to the start of the word following "main."
I agree that it's strange when the start of the next word is the end of the previous word, but the JAccess inspector shows the correct word within this framework.
This behavior can be altered via IDEA settings, specifically under "Settings/Editor/General/When moving by words". For example, you can adjust the settings so that the cursor is always placed at the start of the word. This adjustment will yield the expected behavior, where the inspector displays the current word, including the current character."

**Comment by thgcode** [comment]
I've also reproduced this bug, but I've noticed when selecting text using the control shift arrow keys, NVDA identifies the word boundaries correctly, there might be a difference between the control vs control shift arrow keys handling, I'll check the code to see if I discover anything...

**Comment by thgcode** [comment]
Here's what I discovered so far: it seems the bug is inside the text info, NVDA is constructing its own representation of the text on the [JABTextInfo class](https://github.com/nvaccess/nvda/blob/bf96860d561aa6c1e03f0cbd286be9b04587cb92/source/NVDAObjects/JAB/__init__.py#LL123) based on offsets and when expanding by word it uses the offsets it calculated instead of getting from the API [Code of the caret post moved script](https://github.com/nvaccess/nvda/blob/bf96860d561aa6c1e03f0cbd286be9b04587cb92/source/editableText.py#L148), I checked the code of the accessibility inspector to see how it is displaying the value for the word and it is getting from the getAccessibilityTextItems API. Strangely I found this function defined on the JABHandler.py file but it seems NVDA doesn't use this API at the moment...

**Comment by dmitrii-drobotov** [comment]
I've experimented with implementing the `JABTextInfo._getWordOffsets` method by getting word bounds directly from the JAB API (using `JABContext.getAccessibleTextItems` method), and it seems to solve the issue quite well. Here's my implementation:
https://github.com/nvaccess/nvda/commit/0893c6fc2cbf1130b693695c14a81b3a871b1066.

Words are now read separately as expected, and it also works with different caret behaviors, including the one in IntelliJ where the caret moves to the end of the current word instead of the start of the next word. This is not the final version, as I haven't tested all the edge cases, and also haven't considered the performance yet, but the general approach should stay the same.
I would appreciate if you could test this and confirm if this approach looks right. I can then make a pull request with the fix.

**Comment by kranid** [comment]
@dmitrii-drobotov This implementation works very well when the caret is set to the first or last symbol of the word. However, if the user sets the NVDA review cursor on another symbol and presses CTRL+NVDA+. (laptop keyboard) to speak the current word, this code will call the old implementation to handle this case, and NVDA will speak the incorrect word.

**Comment by dmitrii-drobotov** [comment]
@kranid Thanks for testing, I can also reproduce this issue. I guess we have to go through all possible word locations near the offset, something like the following:
```
	def _getWordOffsets(self, offset):
		word = self.obj.jabContext.getAccessibleTextItems(offset).word
		wordLen = len(word)
		for i in range(0, wordLen + 1):
			if self.obj.jabContext.getAccessibleTextRange(offset - i, offset + wordLen - i - 1) == word:
				return offset - i, offset + wordLen - i

		return super(JABTextInfo, self)._getWordOffsets(offset)
```

Could please check this implementation?

**Comment by kranid** [comment]
@dmitrii-drobotov "I did not discover any real problems with this implementation.
I'm a bit uneasy about having to find the offset of a word despite already having the word and wanting to announce this word. So, I `tried` to find a way to avoid these redundant actions, but all I could come up with was something like this:
```python
	def _getWordOffsets(self, offset):
		word = self.obj.jabContext.getAccessibleTextItems(offset).word
		self.text=word
		wordLen = len(word)
		return 0,wordLen
```

**Comment by mrgrieves4** [comment]
Sorry to hassle, but has this issue been put on hold or is it still being worked on? I'm finding it really hard to use PyCharm with NVDA in its current state and a fix would make a huge difference.

**Comment by kranid** [comment]
@dmitrii-drobotov I have discovered that my code does not work correctly. We cannot set self.text = word because NVDA requires all current text to be available to the text review cursor. Therefore, your implementation is the best approach.

**Comment by dmitrii-drobotov** [comment]
@kranid thanks for looking into this! I also tried to find a more efficient way to implement it but couldn't find any, so I'm thinking of submitting a PR with this implementation.

@seanbudd I see the issue has blocked/needs-product-decision and needs-triage labels. Is there anything I can do to move the issue forward so I could submit a fix? I can clarify the behavior with the proposed solution if needed.

**Comment by seanbudd** [comment]
@dmitrii-drobotov - yes a brief summary of the proposed behaviour and technical implementation would be great

**Comment by dmitrii-drobotov** [comment]
@seanbudd Sure, so the problem is related to moving word by word in a text using Ctrl+Left/Right arrows or NVDA text review commands like move to the next/previous word. Currently, for Java apps NVDA uses a common way to detect word boundaries from `OffsetsTextInfo._getWordOffsets` because `JABTextInfo` doesn't override this method. This default implementation is fine for normal texts, but is not precise with code, and for example, treats a text like "System.out.println" as a single word. So when using Ctrl+Right the cursor will stop at every dot, but NVDA will repeat "System.out.print" multiple times because it treats it as a single word.

The proposed solution is to override the `_getWordOffsets` method in `JABTextInfo` and get the word offsets directly from the Java app, as also suggested in [here in the code comment](https://github.com/nvaccess/nvda/blob/67f6cb99c1484c28ae41ab708be803cb335f9a05/source/textInfos/offsets.py#L149). In general, it works well and solves the problem, but there are a couple of risks:
1. Java Access Bridge doesn't have the exact API to get start and end offsets of the current word (perhaps this is why it was not implemented initially). But there is a close one which returns the word at the cursor's offset as a string. Knowing the word, we can determine its offsets by searching in the range between cursor offset - word length and cursor offset + word length. This looks inefficient, but I don't see a better way to get word offsets. Here's the code for reference:
```
def _getWordOffsets(self, offset):
	word = self.obj.jabContext.getAccessibleTextItems(offset).word
	wordLen = len(word)
	for i in range(0, wordLen + 1):
		if self.obj.jabContext.getAccessibleTextRange(offset - i, offset + wordLen - i - 1) == word:
			return offset - i, offset + wordLen - i

	return super(JABTextInfo, self)._getWordOffsets(offset)
```

2. In IntelliJ-based IDEs specifically, caret movement by word can be configured, and the default way is set to "jump to the current word boundaries", which basically means that Ctrl+Right will move to the end of the current word, rather than to the start of the next word. For example, if I use Ctrl+Right in the text "System.out.println", the cursor will move before the first dot and NVDA will read "System" because IntelliJ returns "System" as the current word. So it might work somewhat unconventionally, but it's definitely an improvement from the current behavior.
Also, IntelliJ has a way to configure caret movement to jump to the word start, which is how it works in pretty much any other Windows app. In this case NVDA will read words in IntelliJ the same as in other apps (Notepad, Chrome, VS Code, etc.).

**Comment by seanbudd** [comment]
Is there a way to do this for just IntelliJ IDE's? I'm worried about wider impact in Java applications if the offset fetching is implemented incorrectly in custom/legacy Java apps

Does the second point mean that implementing this is somewhat unnecessary - i.e. is there a sufficient workaround currently?

**Comment by dmitrii-drobotov** [comment]
@seanbudd 
> Is there a way to do this for just IntelliJ IDE's? I'm worried about wider impact in Java applications if the offset fetching is implemented incorrectly in custom/legacy Java apps

Yes, it looks possible to do if we use the executable name to detect it. Though there will be around 15 different names for every IDE, and it might not detect not well-known forks of the open source IntelliJ Community if they set a custom name.

>Does the second point mean that implementing this is somewhat unnecessary - i.e. is there a sufficient workaround currently?

No, I described the behavior after the fix. Currently it still doesn't detect word boundaries correctly even if the caret movement is configured to be the same as in other apps.

**Comment by seanbudd** [comment]
Thanks for the additional information. I think we can go ahead without the name checking for now, but we can add it later if issues are reported

**Comment by dmitrii-drobotov** [comment]
I've been doing more testing and found that in NetBeans IDE, which isn't related to IntelliJ but also uses Java Access Bridge, there is pretty much the same issue in the main code editor control. However, this fix doesn't resolve it but rather just slightly changes the way punctuation symbols are considered part of the word. I assume it happens because they modified the way the caret moves word by word but didn't reflect the change in accessibility interfaces. I can't say whether the new behavior in NetBeans is better than the current, so I would rather avoid the change there. Probably the same situation can happen in other apps, so I think it's better to limit the change to IntelliJ-based IDEs.

**Comment by kranid** [comment]
@dmitrii-drobotov It seems to me that if NVDA will use the corresponding accessibility interface to retrieve the correct word provided by the applications, the authors of those applications will have the opportunity to change any incorrect behavior. However, if we fix this bug only for IDEA, the authors of NetBeans will not be able to address the current problem with word announcements at all.

**Comment by dmitrii-drobotov** [comment]
@kranid makes sense, but I'm still not sure if it's the right approach as making such changes could be confusing for users. @seanbudd do you have any thoughts on this? I can implement it either way.

To be clear, this is what I get in NetBeans when continuously pressing Ctrl+Right arrow in a line `System.out.println("hello world");` (caret stops marked as | in `System|.|out|.|println|("|Hello |World|");|`) before and after the fix:

Before:

> System.out.println
> System.out.println
> System.out.println
> System.out.println
> ("hello
> ("hello
> world");
> world");
> world");

After:

> System.out.println
> System.out.println
> System.out.println
> System.out.println
> (
> hello
> world
> "
> line feed

So, neither before nor after is correct.

**Comment by seanbudd** [comment]
Yes - we don't want to break behaviour with NetBeans. Unless you can get code to land there which supports this, I think this needs to be limited to java applications. I'd suggest creating an `appModule` with aliases for each application.

**Comment by mltony** [comment]
FWIW an easy workaround to try might be my wordNav add-on - available in add-on store. It solves word navigation problem fundamentally by computing word boundaries on NVDA side and then jumping cursor to the right location in target application. So word definition discrepancy like this would never happen with wordNav.
One potential problem is I believe a user told me a couple of years ago that `TextInfo.updateCaret()` method didn't work for Idea - could be a JAB issue. In this case wordNav won't work since it won't be able to move caret in Idea. But this needs to be confirmed. And hopefully if this is the case, should be easy to fix - hopefully.

**Comment by kranid** [comment]
@seanbudd The fact is that the code is limited to Java applications, but the problem is that NetBeans is also a Java application. Thus, this code will affect word navigation in all Java applications, but proper word navigation will only function correctly in those Java applications whose developers have ensured that the correct words are provided to the accessibility API.
In the current implementation, NVDA attempts to determine word boundaries on its own and often does so incorrectly. This is especially true in IDEs such as IntelliJ IDEA or NetBeans, where in the proposed implementation, word definition is left to the application developers. In IntelliJ IDEA, this will work quite well, as the developers have ensured accurate word definitions. However, in others, such as NetBeans, we cannot predict this, as the developers may not have put in the effort to define words correctly.

**Comment by SaschaCowley** [comment]
Hi @kranid, can you please report this issue to Jetbrains?
